### PR TITLE
fix: support android 15 16k page size for template plugin_ffi 

### PR DESCRIPTION
**Pull request by huanglizhuo**
Android 15 support 16k page sizes, it need [enable 16kb elf alignment](https://developer.android.com/guide/practices/page-sizes#compile-r26-lower).

the current plugin_ffi template will have below error if run on android 15 16k emulator
```
Failed to load dynamic library 'libffigen_app.so': dlopen
failed: empty/missing
DT_HASH/DT_GNU_HASH in
"/data/app/~~Ixsgxu2mj5fKxP1cXpjV6Q==/com.example.ffigen_app_example-6d_efR__WGu
4dsF4tLIaHw==/lib/arm64/libffigen_app.so"
(new hash type from the future?)
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by andrewkolos** [comment]
I don't know much of anything here, but I do wonder if this change could have other implications. Could this break running Flutter and older devices.

Also, even with this PR landed, would the issue reported here persist. Specifically, it would occur instead when we try to load the engine binaries (I don't know what the page size there is).

@reidbaker (cc @dcharkes)

**Comment by reidbaker** [comment]
@zanderso I think you are the most knowledgeable flutter engineer when it comes to 16kb page sizes.

**Comment by reidbaker** [comment]
> Reading up on this, requiring 16kb alignment of elf sections makes sense. I wouldn't worry about that breaking older devices, but it might add some kilobytes to the compiled binary.
> 
> In general, fixing compilation settings in the CMake file for `plugin_ffi` sgtm. Is this a NOOP for other operating systems? The CMake file is shared across all OSes. If not, we'll probably want to guard in the CMake file on target OS.
> 
> Also, we should have a test. Since Android 15 is not yet out, I suppose we don't have devices/simulators in our devicelab @reidbaker?
> 
> And note to self: We also need to fix this in `--template package_ffi` and/or `package:native_toolchain_c`.

Android 15 is out for developers and we have updated the api levels the engine/framework can call to api 35 (android 15). We dont have emulator support in CI yet because, last we checked, launching api 35 emulators was more unstable than our flake budget. 

@dcharkes I am very open to feedback on how we should test. This class of bug stretches what I am comfortable reviewing.

**Comment by reidbaker** [comment]
@dcharkes @zanderso do we also need to build the engine with 16kb page size alignment? 
https://source.android.com/docs/core/architecture/16kb-page-size/16kb

**Comment by dcharkes** [comment]
Maybe we could get our hands on a phone that has an Android 15 beta with the 16kb memory pages and try it locally to see if things work yes or no. If we then verify what changes are needed (for both the engine and plugin_ffi and package_ffi) we could land these without enabling Android 15 tests yet on the CI and leave an open bug for ourselves to enable tests when we deem such infra stable enough. (This is not super satisfactory from a test coverage point of view, but enables us to preempt users running into this being completely not fixed/tested. I presume that once we make the necessary changes to the compiler settings in various places, we're unlikely to break those settings in the meantime.)

I feel fine reviewing an extra compiler flag in a build file after reading what it does, but I would want us to at least locally verify it for ourselves on an Android 15 with 16kb memory pages for ourselves.

**Comment by HosseinYousefi** [comment]
> I feel fine reviewing an extra compiler flag in a build file after reading what it does, but I would want us to at least locally verify it for ourselves on an Android 15 with 16kb memory pages for ourselves.

You can also use an emulator: https://developer.android.com/guide/practices/page-sizes#test

**Comment by reidbaker** [comment]
> Maybe we could get our hands on a phone that has an Android 15 beta with the 16kb memory pages and try it locally to see if things work yes or no. If we then verify what changes are needed (for both the engine and plugin_ffi and package_ffi) we could land these without enabling Android 15 tests yet on the CI and leave an open bug for ourselves to enable tests when we deem such infra stable enough. (This is not super satisfactory from a test coverage point of view, but enables us to preempt users running into this being completely not fixed/tested. I presume that once we make the necessary changes to the compiler settings in various places, we're unlikely to break those settings in the meantime.)
> 
> I feel fine reviewing an extra compiler flag in a build file after reading what it does, but I would want us to at least locally verify it for ourselves on an Android 15 with 16kb memory pages for ourselves.

I am happy to setup a 16kb page emulator and test this pr (if someone has example code ready to compile I would be grateful). I my question around testing was more about what does a unit test or non integration test look like for this class of change.

**Comment by dcharkes** [comment]
For `package_ffi`: `flutter test test/integration.shard/isolated/native_assets_test.dart`.
For `plugin_ffi`: `flutter create --template=plugin_ffi ... && cd xx/example/ && flutter run -d ...` (I don't know which test does this by heart.) @huanglizhuo is this what you did to reproduce it?

Sometimes the emulators don't behave the same way as devices. So I'd only feel comfortable landing a fix if we can first reproduce the error on the emulator without the fix.

**Comment by reidbaker** [comment]
> lutter test test/integration.shard/isolated/native_assets_test.dart

Thank you I am working on getting an emulator setup now with 16kb page sizes.

**Comment by reidbaker** [comment]
running ` flutter test test/integration.shard/isolated/native_assets_test.dart` from `flutter/packages/flutter_tools/` on a tablet setup using the [android 16kb page size instructions](https://developer.android.com/guide/practices/page-sizes#16kb-emulator). 
Confirmed by running `adb shell getconf PAGE_SIZE` returning `16384`. 
All tests pass without this change.

**Comment by reidbaker** [comment]
Using the following commands I tested on several different emulators. 
```
flutter create --template=plugin_ffi android_page_size 
cd android_page_size
flutter create -t plugin_ffi --platforms android android_ffi
cd android_ffi/example
flutter run
```

Flutter red screen with error text listed in this pr description

Flutter red screen with error text listed in this pr description 
* Android 35 tablet with 16kb page size (same emulator that had the integration tests pass) 
* `sdk gphone16k arm64 (mobile)` aka emulated pixel 

Normal template app behavior
Normal template app behavior
* Pixel 8 pro 

Next step is to patch this pr and try again on the failing devices.

**Comment by reidbaker** [comment]
Confirmed that after patching this pr and regenerating a test app all previously broken emulators show the example code correctly. 

```
diff --git a/android_ffi/src/CMakeLists.txt b/android_ffi/src/CMakeLists.txt
index 8867fa5..fb58e86 100644
--- a/android_ffi/src/CMakeLists.txt
+++ b/android_ffi/src/CMakeLists.txt
@@ -15,3 +15,6 @@ set_target_properties(android_ffi PROPERTIES
 )

 target_compile_definitions(android_ffi PUBLIC DART_SHARED_LIB)
+
+# Support android 15 16k page size
+target_link_options(android_ffi PRIVATE "-Wl,-z,max-page-size=16384")
\ No newline at end of file
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 09a3dcc..1d776b3 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -15,3 +15,6 @@ set_target_properties(android_page_size PROPERTIES
 )

 target_compile_definitions(android_page_size PUBLIC DART_SHARED_LIB)
+
+# Support android 15 16k page size
+target_link_options(android_page_size PRIVATE "-Wl,-z,max-page-size=16384")
\ No newline at end of file
```

@dcharkes @HosseinYousefi Do you think a migrator needs to be written for apps that use ffi? 
Also if we can find the test that verifies the plugin_ffi template launched successfully then I would say that this feature will be tested when the android team migrates to api 35 emulators.

**Review by reidbaker — APPROVED** [review]
Thank you for this pr

**Comment by reidbaker** [comment]
flutter/packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart I think is the test that checks plugin_ffi. 

I think a simple test can be written for this by modifying test/commands.shard/permeable/create_test.dart to check for the new flags in the cmake file.

**Comment by reidbaker** [comment]
@huanglizhuo I hope you don't mind I added a test for this change.

**Comment by reidbaker** [comment]
Related to https://github.com/flutter/flutter/issues/149836

**Comment by huanglizhuo** [comment]
> @huanglizhuo I hope you don't mind I added a test for this change.

thank you for the help

**Comment by huanglizhuo** [comment]
> Using the following commands I tested on several different emulators.我使用以下命令在几个不同的模拟器上进行了测试。
> 
> ```
> flutter create --template=plugin_ffi android_page_size 
> cd android_page_size
> flutter create -t plugin_ffi --platforms android android_ffi
> cd android_ffi/example
> flutter run
> ```
> 
> Flutter red screen with error text listed in this pr description
> 
> Flutter red screen with error text listed in this pr description颤动红屏，并显示此 pr 描述中列出的错误文本
> 
> * Android 35 tablet with 16kb page size (same emulator that had the integration tests pass)页面大小为 16kb 的 Android 35 平板电脑（已通过集成测试的同一模拟器）
> * `sdk gphone16k arm64 (mobile)` aka emulated pixel`sdk gphone16k arm64 (mobile)`又名模拟像素
> 
> Normal template app behavior Normal template app behavior正常模板应用程序行为
> 
> * Pixel 8 pro 像素 8 专业版
> 
> Next step is to patch this pr and try again on the failing devices.下一步是修补此 pr 并在出现故障的设备上重试。

sorry for the late response, yes, this the same error I have.

**Comment by dcharkes** [comment]
> flutter/packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart I think is the test that checks plugin_ffi.

Correct. 👍 

> I think a simple test can be written for this by modifying test/commands.shard/permeable/create_test.dart to check for the new flags in the cmake file.

👍 

> @dcharkes @HosseinYousefi Do you think a migrator needs to be written for apps that use ffi?

That could be useful indeed. I believe there's quite some `plugin_ffi` packages out there. I don't know how many of them build from source (as opposed to having a prebuilt dylib). The ones building from source would benefit, prebuilt binaries would not.

We already have a CMake migration here:

* Impl file: lib/src/migrations/cmake_native_assets_migration.dart
* Test dir: test/general.shard/migrations/cmake_project_migration_test.dart

However, that's migrating the flutter app CMake file, not the plugin CMake file.

> running ` flutter test test/integration.shard/isolated/native_assets_test.dart` from `flutter/packages/flutter_tools/` on a tablet setup using the [android 16kb page size instructions](https://developer.android.com/guide/practices/page-sizes#16kb-emulator).
> Confirmed by running `adb shell getconf PAGE_SIZE` returning `16384`.
> All tests pass _without_ this change.

Thanks @reidbaker ! 🙏

**Comment by HosseinYousefi** [comment]
Before merging make sure that this works for Flutter macOS as well as there is no `-z` flag available for `ld` and the cmake file is used for all the OSs not only Android. One way would be to add the target_link_options within an `if (ANDROID)`.

```cmake
if(ANDROID)
  target_link_options({{projectName}} PRIVATE "-Wl,-z,max-page-size=16384")
endif()
```

**Comment by reidbaker** [comment]
> Before merging make sure that this works for Flutter macOS as well as there is no `-z` flag available for `ld` (on mac) and the cmake file is used for all the OSs not only Android. One way would be to add the target_link_options within an `if (ANDROID)`.
> 
> ```cmake
> if(ANDROID)
>   target_link_options({{projectName}} PRIVATE "-Wl,-z,max-page-size=16384")
> endif()
> ```

I can either A) protect the flags with an android check B) wait on you to validate mac, or C) run some validation steps you suggest. 

@HosseinYousefi I think you are near the end of your day. If I dont hear from you then I will just pass these flags for android.

**Comment by HosseinYousefi** [comment]
> > Before merging make sure that this works for Flutter macOS as well as there is no `-z` flag available for `ld` (on mac) and the cmake file is used for all the OSs not only Android. One way would be to add the target_link_options within an `if (ANDROID)`.
> > ```cmake
> > if(ANDROID)
> >   target_link_options({{projectName}} PRIVATE "-Wl,-z,max-page-size=16384")
> > endif()
> > ```
> 
> I can either A) protect the flags with an android check B) wait on you to validate mac, or C) run some validation steps you suggest.
> 
> @HosseinYousefi I think you are near the end of your day. If I dont hear from you then I will just pass these flags for android.

I had to do this migration for `package:jni` and it [did not work on mac](https://github.com/dart-lang/native/actions/runs/11088622285/job/30808769880), so I had to use an [android check](https://github.com/dart-lang/native/blob/main/pkgs/jni/src/CMakeLists.txt#L30). I suggest you do the same here.

**Comment by reidbaker** [comment]
> * Impl file: lib/src/migrations/cmake_native_assets_migration.dart
> * Test dir: test/general.shard/migrations/cmake_project_migration_test.dart

Awesome, thank you for letting me know.  l will make that change now.

**Comment by reidbaker** [comment]
@dcharkes I filed https://github.com/flutter/flutter/issues/155933 to track the migrator work. (and increase the SEO if people hit this error.

**Comment by reidbaker** [comment]
FWIW if either of you know where that exact error message is thrown another potentially easier change would be to mention 16kb page sizes in the error.

**Comment by reidbaker** [comment]
@HosseinYousefi or @dcharkes could one of you double check this pr and give an approval since I have authored a non trivial part of it I feel like my review isnt as valuable.

File: packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl [hunk 1]
```diff
@@ -15,3 +15,8 @@ set_target_properties({{projectName}} PROPERTIES
 )
 
 target_compile_definitions({{projectName}} PUBLIC DART_SHARED_LIB)
+
+if(ANDROID)
+  # Support android 15 16k page size
```
Comment (HosseinYousefi) [inline packages/flutter_tools/templates/plugin_ffi/src.tmpl/CMakeLists.txt.tmpl:20]: nit: Capitalize
```suggestion
  # Support Android 15 16k page size
```

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 2]
```diff
@@ -3149,6 +3149,33 @@ void main() {
...
+    ]);
+
+    final File cmakeLists =
+        globals.fs.file('${projectDir.path}/src/CMakeLists.txt');
+
+    expect(cmakeLists.existsSync(), true);
+
+    final String cmakeListsContent = await cmakeLists.readAsString();
+    const String expected16KbFlags = 'PRIVATE "-Wl,-z,max-page-size=16384")';
+    expect(cmakeListsContent.contains(expected16KbFlags), true);
```
Comment (HosseinYousefi) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:3176]: nit:
```suggestion
    expect(cmakeListsContent, contains(expected16KbFlags));
```

**Comment by zanderso** [comment]
libflutter.so is linked with `-Wl,-z,max-page-size=65536`, which was updated by @rmacnak-google in https://github.com/flutter/buildroot/pull/753.

https://github.com/flutter/buildroot/blob/fb077af8f70dbe087f70497f103b601d27d8a8a8/build/config/compiler/BUILD.gn#L403

**Review by HosseinYousefi — APPROVED** [review]
LGTM! Thanks @huanglizhuo and @reidbaker!

**Comment by reidbaker** [comment]
> libflutter.so is linked with `-Wl,-z,max-page-size=65536`, which was updated by @rmacnak-google in [flutter/buildroot#753](https://github.com/flutter/buildroot/pull/753).
> 
> https://github.com/flutter/buildroot/blob/fb077af8f70dbe087f70497f103b601d27d8a8a8/build/config/compiler/BUILD.gn#L403

Should there be some breadcrumbs for each of these locations. On one hand we dont need to update c compiler flags often but on the other there are at least 3 different places that need compiler flags if android requires them.

**Comment by zanderso** [comment]
> Should there be some breadcrumbs for each of these locations. On one hand we dont need to update c compiler flags often but on the other there are at least 3 different places that need compiler flags if android requires them.

That sounds like a good idea!

**Pull request merged**